source/sundials/arkode.cc

```diff
@@ -302,33 +294,43 @@ namespace SUNDIALS
         yy        = N_VNew_Serial(system_size);
         abs_tolls = N_VNew_Serial(system_size);
       }
-    reset(data.initial_time, data.initial_step_size, solution);
 
-    double next_time = data.initial_time;
+    DiscreteTime time(data.initial_time,
+                      data.final_time,
+                      data.initial_step_size);
+    reset(time.get_current_time(), time.get_next_step_size(), solution);
 
     if (output_step)
-      output_step(0, solution, 0);
+      output_step(time.get_current_time(), solution, time.get_step_number());
 
-    while (t < data.final_time)
+    while (time.is_at_end() == false)
       {
-        next_time += data.output_period;
+        time.set_desired_next_step_size(data.output_period);
 
-        status = SundialsARKode(arkode_mem, next_time, yy, &t, ARK_NORMAL);
+        {
+          double     t;
+          const auto status = SundialsARKode(
+            arkode_mem, time.get_next_time(), yy, &t, ARK_NORMAL);
+          (void)status;
+          AssertARKode(status);
+          time.set_desired_next_step_size(t - time.get_current_time());
+          Assert(std::abs(time.get_next_time() - t) < 1.e-11,
+                 ExcInternalError());
```

Comment:
I don't think I understand these 3 lines. Can you explain what they do?

Reply:
Specifically, what you're doing here makes sure that the next time around `time.advance_time()` (in your line 321) will move the time to the end time. Or do I misunderstand something?

Reply:
Let me try to explain this better. ARKode may choose to change the time-step. So the second argument provided to `SundialsARKode` is the desired next time, and the 4th argument of the function is the actual next time, set by the function. So we use the value of `t` to guide the `time` object, but we expect the `DiscreteTime` class not to modify the step size further. Hence I added the `Assert` statement.

I understand this part of the code is probably not very readable, so I will rewrite this part.

Moreover, I think I should add a new member function called `DiscreteTime::advance_time_to(double new_time)` which guarantees to jump to `new_time` exactly (without any possibility of modifying `new_time`). This method asserts that current_time < new_time <= end_time. The introduction of this member function will simplify the code of this PR as I can just write `time.advance_time_to(t)`.

Reply:
> I think I should add a new member function called DiscreteTime::advance_time_to(double new_time)

Instead of `advance_time_to()` I can introduce the member function `DiscreteTime::set_next_step_size(double step_size)` which is similar to `set_desired_next_step_size()` but does not modify the step size argument. We should use `set_next_step_size()` here because ARKode automatically adjusts the time steps and takes care of stopping at the end_time. Here we don't DiscreteTime to act smart.